plugins/vbe/fu-plugin-vbe.c

```diff
@@ -0,0 +1,314 @@
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "No valid VBE update mechanism found");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_vbe_startup(FuPlugin *plugin, FuProgress *progress, GError **error)
+{
+	FuPluginData *priv = fu_plugin_get_data(plugin);
+	g_autofree gchar *vbe_dir = NULL;
+	g_autofree gchar *state_dir = NULL;
+	g_autofree gchar *bfname = NULL;
+	gchar *buf = NULL;
+	gsize len;
+	gint ret;
+
+	ret = fit_test();
+	if (ret) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_INTERNAL, "fit_test failed: %d", ret);
+		return FALSE;
+	}
+
+	/* get the VBE directory */
+	state_dir = fu_common_get_path(FU_PATH_KIND_LOCALSTATEDIR_PKG);
+	vbe_dir = g_build_filename(state_dir, "vbe", NULL);
+	priv->vbe_dir = g_steal_pointer(&vbe_dir);
+
+	bfname = g_build_filename(priv->vbe_dir, SYSTEM_DT, NULL);
+	if (!g_file_get_contents(bfname, &buf, &len, error)) {
+		/* check if we have a kernel device tree */
+		g_debug("Cannot find system DT '%s'", bfname);
+
+		/* free the filename so we can reuse it */
+		g_free(bfname);
+
+		/* read in the system info */
+		bfname = g_build_filename(KERNEL_DT, NULL);
+		if (!g_file_get_contents(bfname, &buf, &len, error)) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "No kernel device tree '%s'",
+				    bfname);
+			return FALSE;
+		}
+	}
+	g_log(G_LOG_DOMAIN, G_LOG_LEVEL_INFO, "Processing system DT '%s'", bfname);
+	priv->fdt = buf;
+	if (!process_system(priv, buf, len, error)) {
+		g_debug("Failed: %s", (*error)->message);
+		/* error is set by process_system() */
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_vbe_coldplug(FuPlugin *plugin, FuProgress *progress, GError **error)
+{
+	FuContext *ctx = fu_plugin_get_context(plugin);
+	FuPluginData *priv = fu_plugin_get_data(plugin);
+	struct FuVbeMethod *meth;
+	GList *entry;
+
+	/* create a driver for each method */
+	for (entry = g_list_first(priv->methods); entry; entry = g_list_next(entry)) {
+		const struct VbeDriver *driver;
+		g_autoptr(FuDevice) dev;
+		const gchar *version;
+
+		meth = entry->data;
+		driver = meth->driver;
+		dev = driver->new_func(ctx, meth->vbe_method, priv->fdt, meth->node);
+		fu_device_set_id(dev, meth->vbe_method);
+
+		fu_device_set_name(dev, driver->name);
+		fu_device_set_vendor(dev, driver->vendor);
+		fu_device_add_guid(dev, driver->guid);
+
+		fu_device_add_vendor_id(FU_DEVICE(dev), driver->vendor_id);
+		fu_device_set_version_format(dev, FWUPD_VERSION_FORMAT_TRIPLET);
+		fu_device_set_version_lowest(dev, driver->version_lowest);
+
+		version = fdt_getprop(priv->fdt, meth->node, "cur-version", NULL);
+		fu_device_set_version(dev, version);
+
+		version = fdt_getprop(priv->fdt, meth->node, "bootloader-version", NULL);
+		fu_device_set_version_bootloader(dev, version);
+		fu_device_add_icon(dev, "computer");
+		fu_device_add_flag(dev, FWUPD_DEVICE_FLAG_UPDATABLE);
+		fu_plugin_device_add(plugin, dev);
+		g_object_ref(dev);
```

Comment:
I don't think you need to bump the refcounf here, it's a memory leak no?

Reply:
I thought that when dev went out of scope it would remove the device we just created? I cannot test now due to my 'progress' issue

Reply:
`fu_plugin_device_add()` takes a ref on the device.